### PR TITLE
DataRepeater and Pagination Improvements

### PR DESCRIPTION
**Pull request by spekary**
- Refactoring DataRepeater for easier implementation and sub-classing
- Refactoring Paginator
- Removing strActionParameter to fix bugs caused by mixup between that and mixActionParameter
- Moving EvaluateTemplate to ControlBase
- Fixing example

In preparation for Bootstrap plugin.

### CODE REVIEW

File: includes/base_controls/QControlProxy.class.php [hunk 1]
```diff
@@ -34,7 +32,7 @@ public function RenderAsEvents($strActionParameter = null, $blnDisplayOutput = t
 			else
 				$this->strTargetControlId = $this->objForm->GenerateControlId();
 				
-			$this->strActionParameter = $strActionParameter;
+			$this->mixActionParameter = $strActionParameter;
```
Comment (vaibhav-kaushal) [inline includes/base_controls/QControlProxy.class.php]: mixActionParameter is not defined in the class. Please define it. Not doing so produces a soft warning in PhpStorm which can be confusing at best and scaring in worse cases. Please add it the way strActionParameter was present.

File: includes/base_controls/QControlProxy.class.php [hunk 2]
```diff
@@ -62,7 +60,7 @@ public function RenderAsHref($strActionParameter = null, $blnDisplayOutput = tru
 			else
 				$this->strTargetControlId = $this->objForm->GenerateControlId();
 			
-			$this->strActionParameter = $strActionParameter;
+			$this->mixActionParameter = $strActionParameter;
```
Comment (vaibhav-kaushal) [inline includes/base_controls/QControlProxy.class.php]: mixActionParameter is not defined in the class (again)

File: includes/base_controls/QDataGridBase.class.php [hunk 3]
```diff
@@ -825,13 +822,13 @@ protected function GetHeaderRowHtml() {
...
-					$this->strActionParameter = $intColumnIndex;
+					$this->mixActionParameter = $intColumnIndex;
 
 					$strToReturn .= sprintf("    <th %s><a id=\"%s\" href=\"%s\" %s%s>%s</a></th>\r\n",
 						$this->objHeaderRowStyle->GetAttributes(),
-						$this->ControlId . "_col_" . $this->strActionParameter,
+						$this->ControlId . "_col_" . $this->mixActionParameter,
                         			QApplication::$RequestUri,
-						$this->prxDatagridSorting->RenderAsEvents($this->strActionParameter, true, $this->ControlId . "_col_" . $this->strActionParameter, false),
+						$this->prxDatagridSorting->RenderAsEvents($this->mixActionParameter, true, $this->ControlId . "_col_" . $this->mixActionParameter, false),
```
Comment (vaibhav-kaushal) [inline includes/base_controls/QDataGridBase.class.php]: mixActionParameter is not defined in the class (again). Please define it.

File: includes/base_controls/QControlProxy.class.php (hunk 2 above)
Comment (spekary) [inline includes/base_controls/QControlProxy.class.php]: Its defined in QControlBase. The local $strActionParameter was redundant and not necessary. I checked in some minor changes to fix some PhpStorm warnings, but I did not see the warning you were talking about saying that it was undefined.

Comment (vaibhav-kaushal) [inline includes/base_controls/QControlProxy.class.php]: Okay. I had not checked out the code. I have been trying to setup the test environment for alpha but I am too busy with my current work. Can you please write a small guide about how to setup the test environment?

Should I just run composer and then overwrite the directory with git clone?

Comment (spekary) [inline includes/base_controls/QControlProxy.class.php]: Just make a new directory under your docroot and follow the alpha3 installation instructions. Composer does its work with git, so there shouldn't be any additional git commands to issue. The only additional thing at the end would be to install the examples database.

**Comment by vaibhav-kaushal** [comment]
I checked this one. It's working (just adding confirmation, since Oleg has already done the job of confirming the pull).

**Comment by olegabr** [comment]
@vaibhav-kaushal  thank you!

**Pull request merged**